genesis/engine/solvers/fem_solver.py

```diff
@@ -758,40 +757,33 @@ def compute_Ap(self):
                 continue
             V = self.elements_i[i_e].V
             B = self.elements_i[i_e].B
-            s = -B[0, :] - B[1, :] - B[2, :]  # s is the negative sum of B rows
+            i_v = self.elements_i[i_e].el2v
+            S = ti.Matrix.zero(gs.ti_float, 4, 3)
+            S[:3, :] = B
+            S[3, :] = -B[0, :] - B[1, :] - B[2, :]
+
+            if ti.static(self._enable_vertex_constraints):
+                for i in ti.static(range(4)):
+                    if self.vertex_constraints.is_constrained[i_v[i], i_b]:
+                        S[i, :] = ti.Vector.zero(gs.ti_float, 3)
```

Comment:
You have multiple duplication of this pattern. Could this be factored in some private ti.func?